hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -571,18 +571,24 @@ TcpClientConnection getOrConnect(@Nonnull Address address) {
             return connection;
         }
 
-        synchronized (resolveAddress(address)) {
-            // this critical section is used for making a single connection
-            // attempt to the given address at a time.
-            connection = getConnection(address);
-            if (connection != null) {
+        InetSocketAddress inetSocketAddress = resolveAddress(address);
```

Comment:
I'm not sure about the speed at which these objects are created. 

The problem with creating (and destroying) objects with a potentially inflated monitor is that it can be a huge performance hog when doing GC. For example our initial future used the same approach; later we replaced it by lock support to get rid of this inflated monitor and performance increased and latencies became a lot better.

For such situations I would suggest using a striped lock.

Apart from that.. I see you are synchronizing while communicating with external systems. You create the connection, which needs to go through a 3 way handshake (and potentially a few extra steps due to TLS) and then you have authentication. Which is another set of remote steps.

If for whatever reason such a remote interaction stalls, this thread stalls and this can lead to serious problems in the system. 

I thin a bettter approach needs to be found than synchronization.

Reply:
I had the striped lock approach in the first commit. See
https://github.com/hazelcast/hazelcast/pull/17239/commits/e9e36fca17a22c6a1e61e355a8d09acf6a93f402#r500794877
This comment was the reason I switched to putIfAbsent 
https://github.com/hazelcast/hazelcast/pull/17239#discussion_r470445160
Not sure which one is more important. Wdyt?

> I'm not sure about the speed at which these objects are created.

These objects will not be created on a stable cluster. They will be created when there is a disconnection, and the client tries to connect back. 

> Also, any other thread that hits the same address, could also block indefinitely.
About accessing remote under lock. All the remote invocations are given up with a timeout. It can block other threads but it will not block indefinitely.  

> I think a better approach needs to be found than synchronization.
I am open to suggestions.
We may allow opening connections concurrently and clean them up later when we detect they are not used. Seems more complicated to me.